Write SteppedProgress JavaScript

rowanmanning opened the pull request:
This adds some fairly minimal JavaScript to allow for users to interact
with stepped progress. The main use case is progressing the steps,
marking the current step as complete and the next step as current.

It's also possible to directly manipulate each step which caters for
moving backwards through the stepped progress. We can provide
convenience methods for this use case later if we need to.

The best way to see how this works is probably to look at the README
changes first. There's also full JSDoc for all public and private
methods which should hopefully make their purpose clear.

rowanmanning committed:
Write SteppedProgress JavaScript

This adds some fairly minimal JavaScript to allow for users to interact
with stepped progress. The main use case is progressing the steps,
marking the current step as complete and the next step as current.

It's also possible to directly manipulate each step which caters for
moving backwards through the stepped progress. We can provide
convenience methods for this use case later if we need to.

The best way to see how this works is probably to look at the README
changes first. There's also full JSDoc for all public and private
methods which should hopefully make their purpose clear.

JakeChampion commented on src/js/stepped-progress-step.js:
```diff
@@ -0,0 +1,219 @@
+	 * @returns {void}
+	 */
+	_clearStateClasses() {
+		this.stepElement.classList.remove(...stateClassNames);
```
IE11 does not support multiple parameters for the `add` and `remove` methods.
Should we add DOMTokenList to the [`browserFeatures` for the component](https://github.com/Financial-Times/o-stepped-progress/blob/master/origami.json#L23)?

notlee replied:
That DOMTokenList polyfill work paying off 😁

rowanmanning replied:
Probably yes, _or_ I need to rework this code a little bit

rowanmanning replied:
I went with reworking this to loop over the classnames

notlee reviewed (approved):
Looks great to me :D There's an opportunity to improve performance by keeping state in JS and updating classes within `requestanimationframe`, but I suspect it's not a big deal for this component.

JakeChampion commented on src/js/stepped-progress.js:
```diff
@@ -1,24 +1,172 @@
+	 * @returns {Boolean} Returns whether all steps are completed.
+	 */
+	isComplete() {
+		return (this.getCompletedSteps().length === this._steps.length);
```
I'd be tempted to use Array.prototype.every instead of using Array.prototype.filter and comparing lengths. The reasons I would use every are: I think it is more readable and simpler to understand as it describes exactly what the intention is; It is faster because it would return `false` as soon as it finds an incomplete step.
```suggestion
		return this._steps.every(step => step.isComplete())
```

JakeChampion commented on src/js/stepped-progress.js:
```diff
@@ -1,24 +1,172 @@
+	 */
+	_constructSteps() {
+		const elements = this.steppedProgressElement.querySelectorAll(`.${classNames.step}`);
+		this._stepElementInstanceMap = new Map([...elements].map(element => {
```
`_stepElementInstanceMap` looks to only be used in this method, is the reason to have this map attached to the instance to make removing the `SteppedProgressStep` instances simpler?

rowanmanning replied:
Ah yes, I intended to use this to make adding/removing steps easier but then never used it. Perhaps it's better to leave it out for now

rowanmanning replied:
I've removed this for now, we can re-add later if we need a quick way to get the step instance that corresponds to an element

rowanmanning committed:
Fix a lint error

rowanmanning committed:
Remove the _stepElementInstanceMap property

We don't need this for now, and we can revert this commit if we ever do
need it later.

JakeChampion committed:
Use Array.prototype.every for speed and clarity

Co-Authored-By: rowanmanning <accounts+github@rowanmanning.com>

rowanmanning committed:
Remove the need for a polyfill for IE10/11

I decided to keep this simple, and possibly the new code is clearer.
